### PR TITLE
Add  contracts & tests

### PR DESCRIPTION
**Pull request by yusufidimaina9989**

### CODE REVIEW

**Comment by msinkec** [comment]
Good work 👍. 

A few improvements for the crowdfund example:

- You have a global `is_refunded` flag. As I understand this would only allow one donator to issue a refund. There needs to be a separate flag for each donator.
- Rename to something like `CrowdfundStateful` instead of `Crowdfund2`. Also update filename.
- In the `collect` method, use `this.donation` instead of `this.ctx.utxo.value` for clarity. Also build payment output and change output separately, just like in `donate`.
- Make next contract instance be the first output if permitted. (`refund`).
- Use camel case for variables to be consistent with other code.

**Comment by yusufidimaina9989** [comment]
Ok,
thanks

**Comment by yusufidimaina9989** [comment]
Done  0465821ad1cd543f852ef177c5c3f4b181ba01ef

@msinkec

**Comment by msinkec** [comment]
Just noticed the test invocation was not correct. The tests for the crowdsource contract didn't run at all. In the newest commit I set them up for you to finish. The first tests already executes correctly but the next two need to be extended with [a custom call tx builder ](https://docs.scrypt.io/how-to-deploy-and-call-a-contract/how-to-customize-a-contract-tx#call-tx).

I also noticed you already wrote some other tests this way. Could you update those as well?

I also tidied up the contract code a bit and fixed some small issues. See the diff of my commit.

**Comment by yusufidimaina9989** [comment]
Thank you

File: src/contracts/CrowdfundStateful.ts [hunk 1]
```diff
@@ -0,0 +1,171 @@
...
+
+        // refund output
+        const refund = Utils.buildPublicKeyHashOutput(
+            hash160(contributor),
+            amount
+        )
+
+        let outputs = output + refund
+
+        if (this.changeAmount > 0n) {
```
Comment (xhliu) [inline src/contracts/CrowdfundStateful.ts:162]: This is redundant. `buildChangeOutput()` checks this internally already.

Comment (xhliu) [reply src/contracts/CrowdfundStateful.ts:162]: @yusufidimaina9989

File: src/contracts/CrowdfundStateful.ts [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+            'already refunded'
+        )
+
+        // update state
+        const output: ByteString = this.buildStateOutput(
+            this.ctx.utxo.value - amount
+        )
+
+        // setting it as refunded
+        this.donorRefunded.set(contributor, true)
```
Comment (gitzhou) [inline src/contracts/CrowdfundStateful.ts:149]: move this line before `this.buildStateOutput`

it's essential to finish updating all the state props before calling `this.buildStateOutput`

Comment (yusufidimaina9989) [reply src/contracts/CrowdfundStateful.ts:149]: Ok will do that

**Comment by gitzhou** [comment]
btw

- you should NOT add timelock check in `donate` method, people can donate without a cooldown, as well as `refund`, please make sure not add it in these 2 methods
- add timelock check in `collect`, the beneficiary cannot collect before fundraising expired
- add customized tx builder for `collect` and `refund`

**Comment by yusufidimaina9989** [comment]
@msinkec @gitzhou 
plz help me check refund custom tx, still getting `HashOutputs Mismatch` error

`    // User defined transaction builder for calling function `refund`
      static refundTxBuilder(
        current: CrowdfundStateful,
        options: MethodCallOptions<CrowdfundStateful>,
        contributor: PubKey,
        amount: bigint
    ): Promise<ContractTransaction> {
        const nextInstance = current.next()
        const next = options.next as StatefulNext<CrowdfundStateful>

        const unsignedTx: bsv.Transaction = new bsv.Transaction()
            // add contract input
            .addInput(current.buildContractInput())
            // build next instance output
            .addOutput(
                new bsv.Transaction.Output({
                    script: next.instance.lockingScript,
                    satoshis: Number(amount),
                })
            )
            // build refund output
            .addOutput(
                new bsv.Transaction.Output({
                    script: bsv.Script.fromHex(
                        Utils.buildPublicKeyHashScript(
                            hash160(contributor)
                        )
                    ),
                    satoshis: Number(amount),
                })
            )
            // build change output
            if (options.changeAddress){
            unsignedTx.change(options.changeAddress)
            }
        return Promise.resolve({
            tx: unsignedTx,
            atInputIndex: 0,
            nexts: [
                {
                    instance: next.instance,
                    atOutputIndex: 0,
                    balance: Number(amount),
                },
            ],
        })
    }
`

**Comment by gitzhou** [comment]
> .addOutput(
>                 new bsv.Transaction.Output({
>                     script: next.instance.lockingScript,
>                     satoshis: Number(amount),
>                 })
>             )

the amount of contract output (output #0) should be `current.balance - Number(amount)`

**Comment by yusufidimaina9989** [comment]
thank you sir

**Comment by yusufidimaina9989** [comment]
Annotation 2023-09-15 123347

@gitzhou  @msinkec  i'm sorry for disturbing you.
still getting this error please help me out.

**Comment by gitzhou** [comment]
I did an implementation of `crowdfund`, you can refer to it.

https://github.com/sCrypt-Inc/boilerplate/blob/master/src/contracts/crowdfundReplay.ts

**Comment by yusufidimaina9989** [comment]
yes, thanks i will check it out

**Pull request merged**